File: src/main/java/im/joker/helper/LongPollingHelper.java
package im.joker.helper;

import im.joker.api.vo.sync.SyncResponse;
import lombok.extern.slf4j.Slf4j;
import org.redisson.api.RedissonReactiveClient;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import reactor.core.publisher.Mono;
import reactor.core.publisher.MonoSink;

import javax.annotation.PostConstruct;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

import static im.joker.constants.ImContents.SYNC_DEVICE_TOPIC;

@Component
@Slf4j
public class LongPollingHelper {


    @Autowired
    private RedissonReactiveClient redissonReactiveClient;

    /**
     * sync请求等待响应列表,key是DeviceId, value是monoSink
     */
    private final Map<String, MonoSink<Boolean>> waitingDeviceMap = new ConcurrentHashMap<>();

    @PostConstruct
    public void subscribeNotifySyncMessage() {
        redissonReactiveClient.getTopic(SYNC_DEVICE_TOPIC)
                .addListener(String.class, (channel, deviceId) -> {
                    log.debug("收到channel:{},deviceId:{} sync完成的消息", channel, deviceId);
                    if (waitingDeviceMap.containsKey(deviceId)) {
                        MonoSink<Boolean> monoSink = waitingDeviceMap.get(deviceId);
                        monoSink.success(true);
                    }
                }).subscribe();
    }

    /**
     * 唤醒正在同步的长轮询设备
     *
     * @param deviceId
     */
    public Mono<Long> notifySyncDevice(String deviceId) {
        return redissonReactiveClient.getTopic(SYNC_DEVICE_TOPIC).publish(deviceId);
    }

    /**
     * 长沦陷过程中需要添加同步设备,以阻塞
     *
     * @param deviceId
     * @param monoSink
     */

    public void addSyncDevice(String deviceId, MonoSink<Boolean> monoSink) {
        waitingDeviceMap.put(deviceId, monoSink);
    }


    /**
     * 删除对应的monoSink
     *
     * @param deviceId
     */
    public void removeSyncDevice(String deviceId) {
        waitingDeviceMap.remove(deviceId);
    }


}

File: src/main/kotlin/im/joker/helper/LongPollingHelper.kt
package im.joker.helper

import im.joker.constants.ImConstants.Companion.ROOM_SYNC_DEVICE
import im.joker.device.Device
import im.joker.event.room.AbstractRoomEvent
import im.joker.event.room.other.FullReadMarkerEvent
import im.joker.event.room.other.ReceiptEvent
import im.joker.event.room.other.TypingEvent
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.reactive.awaitSingleOrNull
import org.redisson.api.RedissonReactiveClient
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import java.util.concurrent.ConcurrentHashMap
import javax.annotation.PostConstruct

@Component
class LongPollingHelper {

    private val log: Logger = LoggerFactory.getLogger(LongPollingHelper::javaClass.name)


    @Autowired
    private lateinit var redissonReactiveClient: RedissonReactiveClient


    private val waitingSyncMap = ConcurrentHashMap<String, Channel<Boolean>>()

    @Autowired
    private lateinit var roomSubscribeManager: RoomSubscribeManager


    @PostConstruct
    fun init() {
        redissonReactiveClient.getTopic(ROOM_SYNC_DEVICE).addListener(String::class.java) { _, deviceId ->
            log.debug("唤醒device:{}", deviceId)
            waitingSyncMap[deviceId]?.offer(true)
            removeWaitingDevice(deviceId)
        }.subscribe()
    }


    fun addWaitingDevice(deviceId: String, channel: Channel<Boolean>) {
        waitingSyncMap[deviceId] = channel
    }

    suspend fun notifySyncDevice(event: AbstractRoomEvent, device: Device): Unit = coroutineScope {
        val deviceIds = roomSubscribeManager.searchRoomSubscriber(event.roomId)
                .filter {
                    // 如果是打字事件,回执事件不唤醒自己
                    if (event is TypingEvent || event is ReceiptEvent || event is FullReadMarkerEvent) it != device.deviceId else true
                }

        val deferredList = deviceIds
                .map {
                    async { redissonReactiveClient.getTopic(ROOM_SYNC_DEVICE).publish(it).awaitSingleOrNull() }
                }
        deferredList.awaitAll()
    }

    fun removeWaitingDevice(deviceId: String) {
        val remove = waitingSyncMap.remove(deviceId)
        remove?.close()
    }

}